cmake/extensions.cmake

```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+# 4. Zephyr linker function
+########################################################
+# 4.1. zephyr_linker*
+#
+# The following methods are for defining linker structure using CMake functions.
+#
+# This allows Zephyr developers to define linker sections and their content and
+# have this configuration rendered into an appropriate linker script based on
+# the toolchain in use.
+# For example:
+# ld linker scripts with GNU ld
+# ARM scatter files with ARM linker.
+#
+# Example usage:
+# zephyr_linker_section(
+#   NAME my_data
+#   VMA  RAM
+#   LMA  FLASH
+# )
+#
+# and to configure special input sections for the section
+# zephyr_linker_section_configure(
+#   SECTION my_data
+#   INPUT   "my_custom_data"
+#   KEEP
+# )
+
+
+# Usage:
+#   zephyr_linker([FORMAT <format>)
+#
+# Zephyr linker general settings.
+# This function specifies general settings for the linker script to be generated.
+#
+# FORMAT <format>: The output format of the linked executable.
+#
+function(zephyr_linker)
+  set(single_args "FORMAT")
+  cmake_parse_arguments(LINKER "" "${single_args}" "" ${ARGN})
+
+  if(LINKER_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "zephyr_linker_memory(${ARGV0} ...) given unknown "
+                        "arguments: ${MEMORY_UNPARSED_ARGUMENTS}"
+    )
+  endif()
+
+  if(DEFINED LINKER_FORMAT)
+    set_property(TARGET linker APPEND PROPERTY FORMAT ${LINKER_FORMAT})
+  endif()
+endfunction()
+
+# Usage:
+#   zephyr_linker_memory(NAME <name> START <address> SIZE <size> FLAGS <flags>)
+#
+# Zephyr linker memory.
+# This function specifies a memory region for the platform in use.
+#
+# Note:
+#   This function should generally be called with values obtained from device
+#   tree or Kconfig.general settings for the linker script to be generated.
+#
+# NAME <name>    : Name of the memory region, for example FLASH.
+# START <address>: Start address of the memory region.
+#                  Start address can be given as decimal or hex value.
+# SIZE <size>    : Size of the memory region.
+#                  Size can be given as decimal value, hex value, or decimal with postfix k or m.
+#                  All the following are valid values:
+#                    1048576, 0x10000, 1024k, 1024K, 1m, and 1M.
+# FLAGS <flags>  : Flags describing properties of the memory region.
+#                  Currently supported:
+#                  r: Read-only region
+#                  w: Read-write region
+#                  x: Executable region
+#                  The flags r and x, or w and x may be combined like: rx, wx.
+function(zephyr_linker_memory)
+  set(single_args "FLAGS;NAME;SIZE;START")
+  cmake_parse_arguments(MEMORY "" "${single_args}" "" ${ARGN})
+
+  if(MEMORY_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "zephyr_linker_memory(${ARGV0} ...) given unknown "
+                        "arguments: ${MEMORY_UNPARSED_ARGUMENTS}"
+    )
+  endif()
+
+  zephyr_linker_property_append("MEMORY" "${single_args}")
```

Comment:
It depends on where we actually want to control this part.

If we obtain the knowledge of those memory regions from devicetree or Kconfig, should this be ensured there ?
and could we then get into that situation at all.

Or is the CMake part here just passing this on to the script that actually generates the linker, so that script decides if the field is valid or not.

For now, I have tried to minimize the logic in the linker functions in CMake on what is valid values, but exactly on the memory it might make sense to do some basic checking, as it allows us to provide good error messages at CMake configure time, and thus make debugging earlier.

Reply:
> If we obtain the knowledge of those memory regions from devicetree or Kconfig, should this be ensured there ?
> and could we then get into that situation at all.

My opinion: the variable name MEMORY is pretty generic and we shouldn't assume it's not set.